### PR TITLE
 core: clean up memory management for headers

### PR DESCRIPTION
**Pull request by junr03**
Description: this PR completes memory management for headers being passed from the platform to C to Cpp and back.
Risk Level: med - changes memory management of headers.
Testing: Added unit tests to make sure that headers are being freed when appropriate.

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

**Comment by junr03** [comment]
ready for you to take a look @goaway

File: library/common/include/c_types.h [hunk 1]
```diff
@@ -80,11 +80,35 @@ typedef struct {
   envoy_header* headers;
 } envoy_headers;
 
+#ifdef __cplusplus
+extern "C" { // utility functions
+#endif
```
Comment (goaway) [inline library/common/include/c_types.h]: If this is a C implementation, we should drop the extern "C". :)

Comment (goaway) [reply library/common/include/c_types.h]: Nevermind, I got it backwards.

File: library/common/http/header_utility.cc [hunk 2]
```diff
@@ -6,26 +6,24 @@ namespace Envoy {
...
+  if (headers.length > 0) {
+    release_envoy_headers(headers);
+  }
   return transformed_headers;
 }
 
-envoy_headers transformHeaders(const HeaderMap& header_map) {
+envoy_headers toBridgeHeaders(const HeaderMap& header_map) {
   // TODO: provide utility for the caller to free allocated memory
   // https://github.com/lyft/envoy-mobile/issues/280
```
Comment (goaway) [inline library/common/http/header_utility.cc]: I think this is done.

File: library/common/include/c_types.c [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#include "library/common/include/c_types.h"
+
+#include <string.h>
+
+// NOLINT(namespace-envoy)
+// Although this file is in an envoy_cc_library rule it should be C compatible with old style casts.
+#pragma GCC diagnostic ignored "-Wold-style-cast"
```
Comment (goaway) [inline library/common/include/c_types.c]: Might not need the pragma or the NOLINT anymore.

File: test/common/http/header_utility_test.cc [hunk 4]
```diff
@@ -7,18 +7,22 @@
...
 
-envoy_data envoyString(std::string& s) {
-  return {s.size(), reinterpret_cast<const uint8_t*>(s.c_str()), envoy_noop_release, nullptr};
+void envoy_test_release(void* context) {
+  uint32_t* counter = static_cast<uint32_t*>(context);
+  *counter = *counter + 1;
+}
+
+envoy_data envoyTestString(std::string& s, uint32_t* sentinel) {
+  return {s.size(), reinterpret_cast<const uint8_t*>(s.c_str()), envoy_test_release, sentinel};
```
Comment (goaway) [inline test/common/http/header_utility_test.cc]: Down the road, we could consider creating a shared utility for this, if needed.

**Pull request merged**